ReactAndroid/src/main/java/com/facebook/react/uimanager/AccessibilityDelegateUtil.java

```diff
@@ -108,17 +123,34 @@ public static void setDelegate(final View view) {
     // if a view already has an accessibility delegate, replacing it could cause
     // problems,
     // so leave it alone.
-    if (!ViewCompat.hasAccessibilityDelegate(view)
-        && (accessibilityRole != null || view.getTag(R.id.accessibility_states) != null)) {
+    if (!ViewCompat.hasAccessibilityDelegate(view)) {
       ViewCompat.setAccessibilityDelegate(view, new AccessibilityDelegateCompat() {
         @Override
         public void onInitializeAccessibilityNodeInfo(View host, AccessibilityNodeInfoCompat info) {
           super.onInitializeAccessibilityNodeInfo(host, info);
-          setRole(info, accessibilityRole, view.getContext());
-          // states are changable.
-          ReadableArray accessibilityStates = (ReadableArray) view.getTag(R.id.accessibility_states);
-          if (accessibilityStates != null) {
-            setState(info, accessibilityStates, view.getContext());
+
+          if (accessibilityRole != null || view.getTag(R.id.accessibility_states) != null) {
+            setRole(info, accessibilityRole, view.getContext());
+            // states are changable.
+            ReadableArray accessibilityStates = (ReadableArray) view.getTag(R.id.accessibility_states);
+            if (accessibilityStates != null) {
+              setState(info, accessibilityStates, view.getContext());
+            }
+          }
+        }
+
+        @Override
+        public void onInitializeAccessibilityEvent(View host, AccessibilityEvent accessibilityEvent) {
+          super.onInitializeAccessibilityEvent(host, accessibilityEvent);
+          ReactContext reactContext = (ReactContext)host.getContext();
+
+          WritableMap event = Arguments.createMap();
+          event.putInt("target", host.getId());
+
+          if (accessibilityEvent.getEventType() == AccessibilityEvent.TYPE_VIEW_ACCESSIBILITY_FOCUS_CLEARED) {
+            reactContext.getJSModule(RCTEventEmitter.class).receiveEvent(host.getId(), BLUR_EVENT_NAME, event);
```

Comment:
I'm not sure how React Native handles the call to receiveEvent here, but it's worth pointing out that not all events that get passed through `onInitializeAccessibilityEvent` will actually be fired.  Every ancestor of the view has the chance to cancel the event before it bubbles to the top.  Only if it makes it all the way to the top of the view hierarchy will it be passed off to a running accessibility service.

The AccessibilityEvent cycle goes like this:

1.) `View.sendAccessibilityEvent()` (intent to send event started)
2.) `View.onInitializeAccessibilityEvent()` (event populated with view info)
3.) `View.dispatchPopulateAccessibilityEvent()` (event text populated with additional info)
4.) `ViewParent.requestSendAccessibilityEvent()` (parent of View is free to modify or cancel the event, or pass it up to its parent)
5.) ViewRootImpl calls `sendAccessibilityEvent` on the `AccessibilityManager` (this is what actually passes the event to any running accessibility service).

The best way to actually handle this would probably be to override `performAccessibilityAction` instead.  When Talkback (or any other accessibility service) receives a focus event, it's not guaranteed to actually change focus.  For example, if the already focused view sent another `TYPE_VIEW_ACCESSIBILITY_FOCUSED` event, it would be ignored.  If it *does* decide to change focus, it will fire the `performAccessibilityAction` method on the View and pass in an `AccessibilityNodeInfo.ACTION_ACCESSIBILITY_FOCUS` action.

Here is the code in Talkback doing that for reference:
https://github.com/google/talkback/blob/e69d4731fce02bb9e69613d0e48c29033cad4a98/talkback/src/main/java/eventprocessor/ProcessorFocusAndSingleTap.java#L871

Reply:
@bravalla Thanks for this detailed feedback.

I applied your suggestions, which works fine for focus event, but for the blur event it doesn't work as expected. Using `performAccessibilityAction`, blur is triggered just when the focus moves to a non react-native element (**my assumption**). When focus is moving around react-native elements, nothing happens.

In the video you can understand better what I say: blur just trigger when focus moves to status bar.
<img src=https://user-images.githubusercontent.com/20709038/56940068-25235c80-6ada-11e9-9e53-df77fedf9ce8.gif width=300>

code: https://gist.github.com/elucaswork/242c098a78577ddcdef4d75a60d2a4a5

Do you have an idea of what can be happening?
Thank you.

Reply:
@elucaswork, The code here looks good to me.  I've verified the approach in a non-ReactNative app to make sure that Talkback was correctly calling performAccessibilityAction on both focus and blur, and it is.

It's possible that the issue is at the ReactNative side (I commented above on a few typos), but I am definitely not an expert on that layer so I'll defer to others if thats where the issue lies.

You could verify that the Android side is working correctly by adding some logging into the performAccessibilityAction method itself to see if the expected actions are ever being passed in.